From fdb51a0e4f9a247ea6e5395215c31af812be2292 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Sat, 22 Nov 2014 23:56:43 -0500
Subject: [md] dm: Use rcu_dereference() for accessing rcu pointer

Message-id: <1416700613-25142-4-git-send-email-snitzer@redhat.com>
Patchwork-id: 100837
O-Subject: [RHEL7.1 PATCH 03/13] dm: Use rcu_dereference() for accessing rcu pointer
Bugzilla: 1165246
RH-Acked-by: Joe Thornber <thornber@redhat.com>

BZ: 1165246

Upstream linux-dm.git commit 33423974bfc1c61193df765078f0466fece7021e
Author: Pranith Kumar <bobby.prani@gmail.com>
Date:   Tue Oct 28 15:09:56 2014 -0700

    dm: Use rcu_dereference() for accessing rcu pointer

    The map field in 'struct mapped_device' is an rcu pointer. Use rcu_dereference()
    while accessing it.

    Signed-off-by: Pranith Kumar <bobby.prani@gmail.com>
    Signed-off-by: Paul E. McKenney <paulmck@linux.vnet.ibm.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm.c b/drivers/md/dm.c
index 121585f..98f12c6 100644
--- a/drivers/md/dm.c
+++ b/drivers/md/dm.c
@@ -2428,7 +2428,7 @@ static struct dm_table *__bind(struct mapped_device *md, struct dm_table *t,
 
  merge_is_optional = dm_table_merge_is_optional(t);
 
- old_map = md->map;
+ old_map = rcu_dereference(md->map);
  rcu_assign_pointer(md->map, t);
  md->immutable_target_type = dm_table_get_immutable_target_type(t);
 
@@ -2447,7 +2447,7 @@ static struct dm_table *__bind(struct mapped_device *md, struct dm_table *t,
  */
 static struct dm_table *__unbind(struct mapped_device *md)
 {
- struct dm_table *map = md->map;
+ struct dm_table *map = rcu_dereference(md->map);
 
  if (!map)
   return NULL;
@@ -2841,7 +2841,7 @@ int dm_suspend(struct mapped_device *md, unsigned suspend_flags)
   goto out_unlock;
  }
 
- map = md->map;
+ map = rcu_dereference(md->map);
 
  /*
   * DMF_NOFLUSH_SUSPENDING must be set before presuspend.
@@ -2935,7 +2935,7 @@ int dm_resume(struct mapped_device *md)
  if (!dm_suspended_md(md))
   goto out;
 
- map = md->map;
+ map = rcu_dereference(md->map);
  if (!map || !dm_table_get_size(map))
   goto out;
 
-- 
1.7.1